src/main/java/io/supertokens/session/accessToken/AccessTokenSigningKey.java

```diff
@@ -69,142 +73,206 @@ public static AccessTokenSigningKey getInstance(Main main) {
-            this.keyInfo = maybeGenerateNewKeyAndUpdateInDb();
+            this.validKeys = maybeGenerateNewKeyAndUpdateInDb();
         }
-        return this.keyInfo;
+        return this.validKeys;
+    }
+
+    public KeyInfo getCurrentKey()
+            throws StorageQueryException, StorageTransactionLogicException {
+        return this.getKey().get(0);
     }
 
     public synchronized long getKeyExpiryTime() throws StorageQueryException, StorageTransactionLogicException {
         if (Config.getConfig(this.main).getAccessTokenSigningKeyDynamic()) {
             this.getKey();
-            long createdAtTime = this.keyInfo.createdAtTime;
+            // getKey ensures we have at least 1 valid keys
+            long createdAtTime = this.validKeys.get(0).createdAtTime;
             return createdAtTime + Config.getConfig(main).getAccessTokenSigningKeyUpdateInterval();
         } else {
             // return 10 years from now
             return System.currentTimeMillis() + (10L * 365 * 24 * 3600 * 1000);
         }
     }
 
-    private KeyInfo maybeGenerateNewKeyAndUpdateInDb() throws StorageQueryException, StorageTransactionLogicException {
+    private List<KeyInfo> maybeGenerateNewKeyAndUpdateInDb() throws StorageQueryException, StorageTransactionLogicException {
         Storage storage = StorageLayer.getSessionStorage(main);
         CoreConfig config = Config.getConfig(main);
+        
+        final long validityDuration = config.getAccessTokenSigningKeyUpdateInterval() + 2 * config.getAccessTokenValidity();
+        final long keyCurrentAfter = System.currentTimeMillis() - config.getAccessTokenSigningKeyUpdateInterval();
+        final long keysValidAfter = keyCurrentAfter - 2 * config.getAccessTokenValidity();
+        
+        List<KeyInfo> validKeys = null;
 
         if (storage.getType() == STORAGE_TYPE.SQL) {
 
             SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
 
             // start transaction
-            return sqlStorage.startTransaction(con -> {
-                KeyInfo key = null;
-                KeyValueInfo keyFromStorage = sqlStorage.getAccessTokenSigningKey_Transaction(con);
-                if (keyFromStorage != null) {
-                    key = new KeyInfo(keyFromStorage.value, keyFromStorage.createdAtTime);
-                }
-
-                boolean generateNewKey = false;
-
-                if (key != null) {
-                    if (config.getAccessTokenSigningKeyDynamic() && System.currentTimeMillis() > key.createdAtTime
-                            + config.getAccessTokenSigningKeyUpdateInterval()) {
-                        generateNewKey = true;
+            validKeys = sqlStorage.startTransaction(con -> {
+                List<KeyInfo> validKeysFromSQL = new ArrayList<KeyInfo>();
+
+                KeyValueInfo[] keysFromStorage = sqlStorage.getAccessTokenSigningKeys_Transaction(con);
+                
+                boolean generateNewKey = true;
```

Comment:
Maybe you don't need the `generateNewKey` variable because instead of doing `if (generateNewKey) {..}`, you can do `if (validKeysFromSQL.size() == 0) {..}`